# Home work #4 (Spaceship Titanic)

Tasks:
- Считывание датафрейма. Считайте датафрейм и сделайте при считывании все признаки строковыми (hints: используйте dtype).

- Выведите несколько первых, последних, случайных строк (случайные используя встроенный метод и сгенерируйте массив из случайных индексов, по которым будут браться строки)

- Преобразования показателей. Обработайте каждый показатель в соответствии с тем типом данных, которым он должен быть (из первого пункта у нас должны быть изначально все показатели str типа)

- Обработайте пропуски

- Попытайтесь перейти от строкового типа данных к числовым

- Посчитайте базовые статистики показателей используя группировку и встроенные методы

- Попробуйте составить несколько предположений о данных, основываясь на полученных результатах

In [101]:
import pandas as pd

In [102]:
df = pd.read_csv("data/spaceship-titanic/train.csv", sep=',')

In [103]:
df.head()

PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
0     0001_01     Europa     False  B/0/P  TRAPPIST-1e  39.0  False   
1     0002_01      Earth     False  F/0/S  TRAPPIST-1e  24.0  False   
2     0003_01     Europa     False  A/0/S  TRAPPIST-1e  58.0   True   
3     0003_02     Europa     False  A/0/S  TRAPPIST-1e  33.0  False   
4     0004_01      Earth     False  F/1/S  TRAPPIST-1e  16.0  False   

   RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
0          0.0        0.0           0.0     0.0     0.0    Maham Ofracculy   
1        109.0        9.0          25.0   549.0    44.0       Juanna Vines   
2         43.0     3576.0           0.0  6715.0    49.0      Altark Susent   
3          0.0     1283.0         371.0  3329.0   193.0       Solam Susent   
4        303.0       70.0         151.0   565.0     2.0  Willy Santantines   

   Transported  
0        False  
1         True  
2        False  
3        False  
4         True

In [104]:
df.tail()

PassengerId HomePlanet CryoSleep     Cabin    Destination   Age    VIP  \
8688     9276_01     Europa     False    A/98/P    55 Cancri e  41.0   True   
8689     9278_01      Earth      True  G/1499/S  PSO J318.5-22  18.0  False   
8690     9279_01      Earth     False  G/1500/S    TRAPPIST-1e  26.0  False   
8691     9280_01     Europa     False   E/608/S    55 Cancri e  32.0  False   
8692     9280_02     Europa     False   E/608/S    TRAPPIST-1e  44.0  False   

      RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
8688          0.0     6819.0           0.0  1643.0    74.0  Gravior Noxnuther   
8689          0.0        0.0           0.0     0.0     0.0    Kurta Mondalley   
8690          0.0        0.0        1872.0     1.0     0.0       Fayey Connon   
8691          0.0     1049.0           0.0   353.0  3235.0   Celeon Hontichre   
8692        126.0     4688.0           0.0     0.0    12.0   Propsh Hontichre   

      Transported  
8688        False  
8689        False  
8690         True  
8691        False  
8692         True

In [105]:
df.sample(5)

PassengerId HomePlanet CryoSleep     Cabin  Destination   Age    VIP  \
5093     5440_03      Earth     False  F/1041/S  TRAPPIST-1e  27.0  False   
340      0378_05     Europa      True    B/11/S  TRAPPIST-1e  13.0  False   
4833     5156_02     Europa     False   B/200/S  TRAPPIST-1e  29.0  False   
5969     6325_01      Earth      True  G/1026/S  TRAPPIST-1e  34.0  False   
4315     4596_01        NaN      True   G/751/S  TRAPPIST-1e  28.0  False   

      RoomService  FoodCourt  ShoppingMall  Spa  VRDeck             Name  \
5093       1109.0        0.0           0.0  0.0     0.0   Briney Beckeey   
340           0.0        0.0           0.0  0.0     0.0  Polaton Pirejus   
4833          0.0     4198.0         227.0  1.0  1415.0    Atonon Traild   
5969          0.0        0.0           0.0  0.0     0.0    Adamie Benney   
4315          0.0        0.0           0.0  0.0     0.0      Ale Booneid   

      Transported  
5093        False  
340          True  
4833         True  
5969         True  
4315         True

In [109]:
df_rand = df.sample(10)
df_rand

PassengerId HomePlanet CryoSleep     Cabin    Destination   Age    VIP  \
2350     2532_01       Mars     False   F/489/S    TRAPPIST-1e  24.0  False   
3754     4014_02     Europa     False   B/125/P    55 Cancri e  28.0  False   
8212     8776_01      Earth     False   E/574/S    TRAPPIST-1e  40.0  False   
8520     9094_01      Earth      True  G/1465/S  PSO J318.5-22  14.0  False   
134      0146_01      Earth     False    G/25/S    TRAPPIST-1e  56.0  False   
6895     7303_01      Earth     False   E/487/S    TRAPPIST-1e  51.0  False   
6417     6779_01     Europa     False   B/258/S    TRAPPIST-1e  37.0  False   
133      0144_01      Earth     False    F/31/S    TRAPPIST-1e  20.0  False   
4864     5188_03      Earth      True   G/838/P    TRAPPIST-1e  12.0  False   
3947     4213_01      Earth     False   G/691/S  PSO J318.5-22  23.0  False   

     RoomService FoodCourt ShoppingMall    Spa  VRDeck               Name  \
2350       737.0      37.0       1682.0    0.0     0.0         Gorn Traie   
3754         0.0    3238.0          0.0  750.0  2368.0                NaN   
8212         0.0       0.0        286.0    6.0   570.0   Winia Headardyer   
8520         0.0       0.0          0.0    0.0     0.0     Aleen Hayertez   
134          0.0       0.0          0.0    0.0     0.0  Verly Wolferguson   
6895         3.0      33.0         12.0   38.0   896.0   Lonnez Adavisons   
6417         0.0    7920.0        270.0    1.0   322.0      Terope Busive   
133          0.0     420.0        210.0    0.0     0.0      Guadae Dayers   
4864         0.0       0.0          0.0    0.0     0.0       Wally Bootez   
3947         4.0     157.0        142.0    NaN   196.0   Fayen Morrodgers   

     Transported  
2350       False  
3754       False  
8212        True  
8520       False  
134        False  
6895       False  
6417        True  
133         True  
4864       False  
3947       False

In [107]:
df.dtypes

PassengerId      object
HomePlanet       object
CryoSleep        object
Cabin            object
Destination      object
Age             float64
VIP              object
RoomService     float64
FoodCourt       float64
ShoppingMall    float64
Spa             float64
VRDeck          float64
Name             object
Transported        bool
dtype: object

In [108]:
#изменение типа признаков
df = df.astype(dtype=object)
df.dtypes

PassengerId     object
HomePlanet      object
CryoSleep       object
Cabin           object
Destination     object
Age             object
VIP             object
RoomService     object
FoodCourt       object
ShoppingMall    object
Spa             object
VRDeck          object
Name            object
Transported     object
dtype: object

In [111]:
#преобразование показателей
df = df.astype({'Age': float, 'RoomService': float, 'FoodCourt': float, 'ShoppingMall': float, 'Spa': float, 'VRDeck': float})
df.dtypes

PassengerId      object
HomePlanet       object
CryoSleep        object
Cabin            object
Destination      object
Age             float64
VIP              object
RoomService     float64
FoodCourt       float64
ShoppingMall    float64
Spa             float64
VRDeck          float64
Name             object
Transported      object
dtype: object

In [112]:
df = df[df.Age.notna()]
df = df.drop("Cabin", axis=1, errors='ignore')
df = df.drop("Destination", axis=1, errors='ignore')
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8514 entries, 0 to 8692
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   PassengerId   8514 non-null   object 
 1   HomePlanet    8315 non-null   object 
 2   CryoSleep     8298 non-null   object 
 3   Age           8514 non-null   float64
 4   VIP           8317 non-null   object 
 5   RoomService   8334 non-null   float64
 6   FoodCourt     8335 non-null   float64
 7   ShoppingMall  8310 non-null   float64
 8   Spa           8332 non-null   float64
 9   VRDeck        8330 non-null   float64
 10  Name          8317 non-null   object 
 11  Transported   8514 non-null   object 
dtypes: float64(6), object(6)
memory usage: 864.7+ KB


In [113]:
df.HomePlanet.value_counts(dropna=False)

Earth     4514
Europa    2085
Mars      1716
NaN        199
Name: HomePlanet, dtype: int64

In [114]:
df["HomePlanet"] = df["HomePlanet"].fillna('Earth')
df.HomePlanet.value_counts(dropna=False)

Earth     4713
Europa    2085
Mars      1716
Name: HomePlanet, dtype: int64

In [116]:
df.CryoSleep.value_counts(dropna=False)
df["CryoSleep"] = df["CryoSleep"].fillna('False')
df.CryoSleep = df.CryoSleep.replace({False: 0, True: 1})
df.CryoSleep.unique()

array([0, 1], dtype=int64)

In [117]:
df.VIP.value_counts(dropna=False)
df["VIP"] = df["VIP"].fillna(False)
df.VIP = df.VIP.replace({False: 0, True: 1})
df.VIP.unique()

array([0, 1], dtype=int64)

In [119]:
df.Transported = df.Transported.replace({False: 0, True: 1})

In [120]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8514 entries, 0 to 8692
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   PassengerId   8514 non-null   object 
 1   HomePlanet    8514 non-null   object 
 2   CryoSleep     8514 non-null   int64  
 3   Age           8514 non-null   float64
 4   VIP           8514 non-null   int64  
 5   RoomService   8334 non-null   float64
 6   FoodCourt     8335 non-null   float64
 7   ShoppingMall  8310 non-null   float64
 8   Spa           8332 non-null   float64
 9   VRDeck        8330 non-null   float64
 10  Name          8317 non-null   object 
 11  Transported   8514 non-null   int64  
dtypes: float64(6), int64(3), object(3)
memory usage: 864.7+ KB


In [126]:
df[["RoomService", "VIP"]][df.RoomService != ''].groupby("VIP").agg(["mean", "median", "min", "max"])

RoomService                     
           mean median  min      max
VIP                                 
0    220.155283    0.0  0.0  14327.0
1    467.989691    0.0  0.0   8243.0

In [127]:
df = df.fillna({"RoomService": 0.0, "FoodCourt": 0.0, "ShoppingMall": 0.0, "Spa": 0.0, "VRDeck": 0.0, "Name": "NA"})
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8514 entries, 0 to 8692
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   PassengerId   8514 non-null   object 
 1   HomePlanet    8514 non-null   object 
 2   CryoSleep     8514 non-null   int64  
 3   Age           8514 non-null   float64
 4   VIP           8514 non-null   int64  
 5   RoomService   8514 non-null   float64
 6   FoodCourt     8514 non-null   float64
 7   ShoppingMall  8514 non-null   float64
 8   Spa           8514 non-null   float64
 9   VRDeck        8514 non-null   float64
 10  Name          8514 non-null   object 
 11  Transported   8514 non-null   int64  
dtypes: float64(6), int64(3), object(3)
memory usage: 864.7+ KB


In [128]:
df.describe()

CryoSleep          Age          VIP   RoomService     FoodCourt  \
count  8514.000000  8514.000000  8514.000000   8514.000000   8514.000000   
mean      0.347075    28.827930     0.023256    221.147992    449.297980   
std       0.476068    14.489021     0.150724    663.939245   1598.407975   
min       0.000000     0.000000     0.000000      0.000000      0.000000   
25%       0.000000    19.000000     0.000000      0.000000      0.000000   
50%       0.000000    27.000000     0.000000      0.000000      0.000000   
75%       1.000000    38.000000     0.000000     42.000000     62.000000   
max       1.000000    79.000000     1.000000  14327.000000  29813.000000   

       ShoppingMall           Spa        VRDeck  Transported  
count   8514.000000   8514.000000   8514.000000  8514.000000  
mean     171.043810    303.636951    298.762156     0.503641  
std      602.376961   1120.733002   1138.416271     0.500016  
min        0.000000      0.000000      0.000000     0.000000  
25%        0.000000      0.000000      0.000000     0.000000  
50%        0.000000      0.000000      0.000000     1.000000  
75%       23.000000     54.000000     40.000000     1.000000  
max    23492.000000  22408.000000  24133.000000     1.000000

In [130]:
df[["Age", "Transported"]].groupby("Transported").agg(["mean", "median"])

Age       
                  mean median
Transported                  
0            29.922858   27.0
1            27.748834   26.0

In [134]:
df[["Age"]].agg(["mean", "median"])

Age
mean    28.82793
median  27.00000

In [141]:
df.Transported.value_counts(normalize=True)

1    0.503641
0    0.496359
Name: Transported, dtype: float64

In [145]:
df.VIP.value_counts(normalize=True)

0    0.976744
1    0.023256
Name: VIP, dtype: float64

In [135]:
df.HomePlanet.value_counts(normalize=True)

Earth     0.553559
Europa    0.244891
Mars      0.201550
Name: HomePlanet, dtype: float64

In [136]:
df.CryoSleep.value_counts(normalize=True)

0    0.652925
1    0.347075
Name: CryoSleep, dtype: float64

In [147]:
df[["CryoSleep", "Transported"]].groupby("Transported").value_counts(normalize=True)

Transported  CryoSleep
0            0            0.873166
             1            0.126834
1            1            0.564132
             0            0.435868
dtype: float64

In [155]:
df[["HomePlanet", "Transported"]].groupby("HomePlanet").value_counts(normalize=True)

HomePlanet  Transported
Earth       0              0.572035
            1              0.427965
Europa      1              0.658513
            0              0.341487
Mars        1              0.523310
            0              0.476690
dtype: float64

In [156]:
df[["VIP", "Transported"]].groupby("VIP").value_counts(normalize=True)

VIP  Transported
0    1              0.506494
     0              0.493506
1    0              0.616162
     1              0.383838
dtype: float64

Resume:

- около 50% пассажиров перешли в иное измерение, из них 43.6% находилось в креосне (при этом в целом на корабле в креосне находилось только 34.7%). Есть гипотеза, что вероятность транспортировки в иное измерение у пассажиров, находящихся в креосне, выше.

- средний возраст пассажиров - 27 лет, средний возраст перешедших в другое измерение - 26 лет;

- из всех пассажиров с Марса транспортировалось - 52.3%, из всех пассажиров с Европы транспортировалось - 65.8%, из всех пассажиров с Земли транспортировалось - 42.7%. При этом численность всех пассажиров с Земли превышает численность всех пассажиров с Марса и с Европы в два раза по каждой группе в отдельности. Есть гипотеза, что вероятность транспортировки в иное измерение у пассажиров с Земли ниже.